cache/cache.go

```diff
@@ -0,0 +1,496 @@
+		item.expiration = time.Now().UnixNano()
+	}
+	c.mu.Unlock()
+	c.metrics.incCacheRequests(StatusSuccess)
+	return nil
+}
+
+// Clear all index from the cache.
+// This reallocates the underlying array holding the index,
+// so that the memory used by the index is reclaimed.
+// A closed cache cannot be cleared.
+func (c *cache[T]) Clear() {
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		return
+	}
+	c.index = make(map[string]*item[T])
+	c.mu.Unlock()
+}
+
+// ListKeys returns a slice of the keys in the cache.
+// If the cache is closed, ListKeys returns nil.
+func (c *cache[T]) ListKeys() []string {
+	c.mu.RLock()
+	if c.closed {
+		c.mu.RUnlock()
+		c.metrics.incCacheRequests(StatusFailure)
+		return nil
+	}
+	keys := make([]string, 0, len(c.index))
+	for k := range c.index {
+		keys = append(keys, k)
+	}
+	c.mu.RUnlock()
+	c.metrics.incCacheRequests(StatusSuccess)
+	return keys
+}
+
+// Resize resizes the cache and returns the number of index removed.
+func (c *cache[T]) Resize(size int) int {
+	overflow := len(c.items) - size
+	if overflow <= 0 {
+		c.metrics.incCacheRequests(StatusSuccess)
+		return 0
+	}
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		c.metrics.incCacheRequests(StatusFailure)
+		return 0
+	}
+
+	if !c.sorted {
+		// sort the slice of index by expiration time
+		slices.SortFunc(c.items, func(i, j *item[T]) int {
+			return cmp.Compare(i.expiration, j.expiration)
+		})
+		c.sorted = true
+	}
+
+	// delete the overflow indexes
+	for _, v := range c.items[:overflow] {
+		delete(c.index, v.key)
+		c.metrics.incCacheEvictions()
+		c.metrics.decCacheItems()
+	}
+	// remove the overflow indexes from the slice
+	c.items = c.items[overflow:]
+	c.mu.Unlock()
+	c.metrics.incCacheRequests(StatusSuccess)
+	return overflow
+}
+
+// HasExpired returns true if the item has expired.
+func (c *Cache[T]) HasExpired(object T) (bool, error) {
+	key, err := c.keyFunc(object)
+	if err != nil {
+		return false, KeyError{object, err}
+	}
+
+	c.mu.RLock()
+	if c.closed {
+		c.mu.RUnlock()
+		c.metrics.incCacheRequests(StatusFailure)
+		return false, KeyError{object, ErrClosed}
+	}
+	item, ok := c.index[key]
+	if !ok {
+		c.mu.RUnlock()
+		c.metrics.incCacheRequests(StatusSuccess)
+		return true, nil
+	}
+
+	if item.expiration < time.Now().UnixNano() {
+		c.mu.RUnlock()
+		c.metrics.incCacheRequests(StatusSuccess)
+		return true, nil
+	}
+
+	c.mu.RUnlock()
+	c.metrics.incCacheRequests(StatusSuccess)
+	return false, nil
+}
+
+// SetExpiration sets the expiration for the given key.
+// Does nothing if the key is not in the cache.
+func (c *Cache[T]) SetExpiration(object T, expiration time.Duration) error {
+	key, err := c.keyFunc(object)
+	if err != nil {
+		return KeyError{object, err}
+	}
+
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		c.metrics.incCacheRequests(StatusFailure)
+		return KeyError{object, ErrClosed}
+	}
+	item, ok := c.index[key]
+	if ok {
+		item.expiration = time.Now().Add(expiration).UnixNano()
```

Comment:
What is the issue with setting negative expiration? If won't panic. The consequences I see is for it to be expired, and so it will be garbage collected, so that's fine. There can be overflow, but we can't resonnably guard against that.